📝(doc) create complete documentation website for Meet (#1436)
numerique-gouv/meet: docs/self-hosting/configuration/turn.md

# TURN Server

> TURN (Traversal Using Relays around NAT) is a relay protocol that routes WebRTC media through a well-known port when direct UDP is blocked. Without TURN, participants on corporate networks, hotel Wi-Fi, or VPNs may connect to a meeting but have no audio or video.

Reviewer: ```suggestion
TURN (Traversal Using Relays around NAT) is a relay protocol that routes WebRTC media through a well-known port when direct UDP is blocked. Without TURN, participants on corporate networks, hotel Wi-Fi, or VPNs may not connect to a meeting.
```

---

## Do you need TURN?

| Deployment | Recommendation |
|---|---|
| Internal team, controlled network | TURN optional - direct media (7882/UDP) works |
| Public instance, mixed networks | TURN recommended |

> | Enterprise | TURN required - corporate firewalls routinely block non-standard UDP |

Reviewer: think we could be a bit more precise here.

Some corporate firewalls block all UDP traffic, forcing clients to fall back to TURN over TCP (or TLS).

TURN is also required when the client is behind a symmetric NAT, where STUN alone cannot establish a direct peer-to-peer connection. For less restrictive NAT types, STUN is usually sufficient. In fact, a STUN server is almost essential for WebRTC; LiveKit uses Google's public STUN server by default if you don't configure your own.

TURN over UDP may still be required depending on the client's NAT and firewall configuration. In general, LiveKit gathers multiple ICE candidates (host, STUN, TURN over UDP/TCP/TLS) and lets ICE negotiate the best available path.

> **How to tell if TURN is needed:** participants who connect (room appears joined, participants list shows them) but have no audio or video, and this resolves when they switch to a phone hotspot - that is a UDP-blocked network. TURN solves it.

Reviewer: LiveKit is not only about UDP. I think this part of the documentation is a bit oversimplified.

LiveKit offers a range of connection candidates, some over UDP and others over TCP, depending on what is available.

It might be related to how the LiveKit server is configured.

---

## How LiveKit TURN works

LiveKit has a built-in TURN server. When a participant's ICE negotiation fails over UDP 7882, the browser falls back to TURN:

1. Browser connects to LiveKit via WebSocket (443/TCP, already working)
2. LiveKit advertises its TURN address in ICE candidates
3. Browser establishes a TURN allocation on 443/UDP
4. All media is relayed through that allocation

The result: media travels over a port that firewalls universally allow, indistinguishable from regular HTTPS.

---

> ## Option 1: Built-in TURN on 443/UDP (recommended)

Reviewer: One thing that could also be worth mentioning is that LiveKit's built-in TURN server doesn't support every TURN transport (e.g. TURN over UDP, TCP, and TLS). In some environments, especially those with restrictive corporate networks, it can be beneficial to deploy a dedicated TURN server such as coturn to provide the transports you need.
